File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManagerImpl.java
package com.reactnativecommunity.webview;

import androidx.annotation.Nullable;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.common.build.ReactBuildConfig;
import com.facebook.react.uimanager.ThemedReactContext;

import android.app.Activity;
import android.app.DownloadManager;
import android.content.pm.ActivityInfo;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.webkit.CookieManager;
import android.webkit.DownloadListener;
import android.webkit.URLUtil;
import android.webkit.WebSettings;
import android.webkit.WebView;

import java.io.UnsupportedEncodingException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.Locale;

public class RNCWebViewManagerImpl {

    public static final String NAME = "RNCWebView";
    private static final String TAG = "RNCWebViewManagerImpl";

    protected static final String HTML_ENCODING = "UTF-8";
    protected static final String HTML_MIME_TYPE = "text/html";
    protected static final String JAVASCRIPT_INTERFACE = "ReactNativeWebView";
    protected static final String HTTP_METHOD_POST = "POST";
    // Use `webView.loadUrl("about:blank")` to reliably reset the view
    // state and release page resources (including any running JavaScript).
    protected static final String BLANK_URL = "about:blank";

    protected static final String DEFAULT_DOWNLOADING_MESSAGE = "Downloading";
    protected static final String DEFAULT_LACK_PERMISSION_TO_DOWNLOAD_MESSAGE =
            "Cannot download files as permission was denied. Please provide permission to write to storage, in order to download files.";

    public static RNCWebView createViewInstance(ThemedReactContext reactContext, @Nullable RNCWebViewConfig webViewConfig, @Nullable String downloadingMessage, @Nullable String lackPermissionToDownloadMessage, boolean allowsFullscreenVideo) {
        RNCWebView webView = new RNCWebView(reactContext);
        setupWebChromeClient(reactContext, webView, allowsFullscreenVideo);
        reactContext.addLifecycleEventListener(webView);
        webViewConfig.configWebView(webView);
        WebSettings settings = webView.getSettings();
        settings.setBuiltInZoomControls(true);
        settings.setDisplayZoomControls(false);
        settings.setDomStorageEnabled(true);
        settings.setSupportMultipleWindows(true);
    
        settings.setAllowFileAccess(false);
        settings.setAllowContentAccess(false);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
          settings.setAllowFileAccessFromFileURLs(false);
          setAllowUniversalAccessFromFileURLs(webView, false);
        }
        setMixedContentMode(webView, "never");
    
        // Fixes broken full-screen modals/galleries due to body height being 0.
        webView.setLayoutParams(
          new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT));
    
        if (ReactBuildConfig.DEBUG && Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
          WebView.setWebContentsDebuggingEnabled(true);
        }
        // webView.setDownloadListener(new DownloadListener() {
        //     public void onDownloadStart(String url, String userAgent, String contentDisposition, String mimetype, long contentLength) {
        //       webView.setIgnoreErrFailedForThisURL(url);
      
        //       RNCWebViewModule module = getModule(reactContext);
      
        //       DownloadManager.Request request;
        //       try {
        //         request = new DownloadManager.Request(Uri.parse(url));
        //       } catch (IllegalArgumentException e) {
        //         Log.w(TAG, "Unsupported URI, aborting download", e);
        //         return;
        //       }
      
        //       String fileName = URLUtil.guessFileName(url, contentDisposition, mimetype);
        //       String downloadMessage = "Downloading " + fileName;
      
        //       //Attempt to add cookie, if it exists
        //       URL urlObj = null;
        //       try {
        //         urlObj = new URL(url);
        //         String baseUrl = urlObj.getProtocol() + "://" + urlObj.getHost();
        //         String cookie = CookieManager.getInstance().getCookie(baseUrl);
        //         request.addRequestHeader("Cookie", cookie);
        //       } catch (MalformedURLException e) {
        //         Log.w(TAG, "Error getting cookie for DownloadManager", e);
        //       }
      
        //       //Finish setting up request
        //       request.addRequestHeader("User-Agent", userAgent);
        //       request.setTitle(fileName);
        //       request.setDescription(downloadMessage);
        //       request.allowScanningByMediaScanner();
        //       request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);
        //       request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, fileName);
      
        //       module.setDownloadRequest(request);
      
        //       if (module.grantFileDownloaderPermissions(getDownloadingMessageOrDefault(downloadingMessage), getLackPermissionToDownloadMessageOrDefault(lackPermissionToDownloadMessage))) {
        //         module.downloadFile(getDownloadingMessageOrDefault(downloadingMessage));
        //       }
        //     }
        //   });
      
        return webView;
    }

    protected static void setupWebChromeClient(ThemedReactContext reactContext, RNCWebView webView, boolean allowsFullscreenVideo) {
        Activity activity = reactContext.getCurrentActivity();

        if (allowsFullscreenVideo && activity != null) {
            int initialRequestedOrientation = activity.getRequestedOrientation();

            RNCWebChromeClient webChromeClient = new RNCWebChromeClient(reactContext, webView) {
                @Override
                public Bitmap getDefaultVideoPoster() {
                    return Bitmap.createBitmap(50, 50, Bitmap.Config.ARGB_8888);
                }

                @Override
                public void onShowCustomView(View view, CustomViewCallback callback) {
                    if (mVideoView != null) {
                        callback.onCustomViewHidden();
                        return;
                    }

                    mVideoView = view;
                    mCustomViewCallback = callback;

                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED);

                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                        mVideoView.setSystemUiVisibility(FULLSCREEN_SYSTEM_UI_VISIBILITY);
                        activity.getWindow().setFlags(
                                WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
                                WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
                        );
                    }

                    mVideoView.setBackgroundColor(Color.BLACK);

                    // Since RN's Modals interfere with the View hierarchy
                    // we will decide which View to hide if the hierarchy
                    // does not match (i.e., the WebView is within a Modal)
                    // NOTE: We could use `mWebView.getRootView()` instead of `getRootView()`
                    // but that breaks the Modal's styles and layout, so we need this to render
                    // in the main View hierarchy regardless
                    ViewGroup rootView = getRootView();
                    rootView.addView(mVideoView, FULLSCREEN_LAYOUT_PARAMS);

                    // Different root views, we are in a Modal
                    if (rootView.getRootView() != mWebView.getRootView()) {
                        mWebView.getRootView().setVisibility(View.GONE);
                    } else {
                        // Same view hierarchy (no Modal), just hide the WebView then
                        mWebView.setVisibility(View.GONE);
                    }

                    mReactContext.addLifecycleEventListener(this);
                }

                @Override
                public void onHideCustomView() {
                    if (mVideoView == null) {
                        return;
                    }

                    // Same logic as above
                    ViewGroup rootView = getRootView();

                    if (rootView.getRootView() != mWebView.getRootView()) {
                        mWebView.getRootView().setVisibility(View.VISIBLE);
                    } else {
                        // Same view hierarchy (no Modal)
                        mWebView.setVisibility(View.VISIBLE);
                    }

                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                        activity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS);
                    }

                    rootView.removeView(mVideoView);
                    mCustomViewCallback.onCustomViewHidden();

                    mVideoView = null;
                    mCustomViewCallback = null;

                    activity.setRequestedOrientation(initialRequestedOrientation);

                    mReactContext.removeLifecycleEventListener(this);
                }
            };

            webView.setWebChromeClient(webChromeClient);
        } else {
            RNCWebChromeClient webChromeClient = (RNCWebChromeClient) webView.getWebChromeClient();
            if (webChromeClient != null) {
                webChromeClient.onHideCustomView();
            }

            webChromeClient = new RNCWebChromeClient(reactContext, webView) {
                @Override
                public Bitmap getDefaultVideoPoster() {
                    return Bitmap.createBitmap(50, 50, Bitmap.Config.ARGB_8888);
                }
            };

            webView.setWebChromeClient(webChromeClient);
        }
    }

    // public static RNCWebViewModule getModule(ThemedReactContext reactContext) {
    //     return reactContext.getNativeModule(RNCWebViewModule.class);
    // }

    public static void setMixedContentMode(WebView view, @Nullable String mixedContentMode) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            if (mixedContentMode == null || "never".equals(mixedContentMode)) {
                view.getSettings().setMixedContentMode(WebSettings.MIXED_CONTENT_NEVER_ALLOW);
            } else if ("always".equals(mixedContentMode)) {
                view.getSettings().setMixedContentMode(WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);
            } else if ("compatibility".equals(mixedContentMode)) {
                view.getSettings().setMixedContentMode(WebSettings.MIXED_CONTENT_COMPATIBILITY_MODE);
            }
        }
    }

    public static void setAllowUniversalAccessFromFileURLs(WebView view, boolean allow) {
        view.getSettings().setAllowUniversalAccessFromFileURLs(allow);
    }

    private static String getDownloadingMessageOrDefault(@Nullable String downloadingMessage) {
        return downloadingMessage == null ? DEFAULT_DOWNLOADING_MESSAGE : downloadingMessage;
    }

    private static String getLackPermissionToDownloadMessageOrDefault(@Nullable String lackPermissionToDownloadMessage) {
        return lackPermissionToDownloadMessage == null ? DEFAULT_LACK_PERMISSION_TO_DOWNLOAD_MESSAGE : lackPermissionToDownloadMessage;
    }

    public static void setColor(RNCWebView view, String color) {
        view.setBackgroundColor(Color.parseColor(color));
    }

    public static void setSource(RNCWebView view, ReadableMap source) {
        if (source != null) {
            if (source.hasKey("html")) {
                String html = source.getString("html");
                String baseUrl = source.hasKey("baseUrl") ? source.getString("baseUrl") : "";
                view.loadDataWithBaseURL(baseUrl, html, HTML_MIME_TYPE, HTML_ENCODING, null);
                return;
            }
            if (source.hasKey("uri")) {
                String url = source.getString("uri");
                String previousUrl = view.getUrl();
                if (previousUrl != null && previousUrl.equals(url)) {
                    return;
                }
                if (source.hasKey("method")) {
                    String method = source.getString("method");
                    if (method.equalsIgnoreCase(HTTP_METHOD_POST)) {
                        byte[] postData = null;
                        if (source.hasKey("body")) {
                            String body = source.getString("body");
                            try {
                                postData = body.getBytes("UTF-8");
                            } catch (UnsupportedEncodingException e) {
                                postData = body.getBytes();
                            }
                        }
                        if (postData == null) {
                            postData = new byte[0];
                        }
                        view.postUrl(url, postData);
                        return;
                    }
                }
                HashMap<String, String> headerMap = new HashMap<>();
                if (source.hasKey("headers")) {
                    ReadableMap headers = source.getMap("headers");
                    ReadableMapKeySetIterator iter = headers.keySetIterator();
                    while (iter.hasNextKey()) {
                        String key = iter.nextKey();
                        if ("user-agent".equals(key.toLowerCase(Locale.ENGLISH))) {
                            if (view.getSettings() != null) {
                                view.getSettings().setUserAgentString(headers.getString(key));
                            }
                        } else {
                            headerMap.put(key, headers.getString(key));
                        }
                    }
                }
                view.loadUrl(url, headerMap);
                return;
            }
        }
        view.loadUrl(BLANK_URL);
    }
}

File: android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManagerImpl.kt
package com.reactnativecommunity.webview

import android.app.DownloadManager
import android.content.Context
import android.content.pm.ActivityInfo
import android.graphics.Bitmap
import android.graphics.Color
import android.net.Uri
import android.os.Environment
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.webkit.*
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.common.build.ReactBuildConfig
import java.io.UnsupportedEncodingException
import java.net.MalformedURLException
import java.net.URL
import java.util.*


class RNCWebViewManagerImpl(context: ReactApplicationContext) {
    companion object {
        const val NAME = "RNCWebView"
    }

    private val TAG = "RNCWebViewManagerImpl"
    private var mWebViewConfig: RNCWebViewConfig = RNCWebViewConfig { webView: WebView? -> }
    private var mAllowsFullscreenVideo = false
    private var mDownloadingMessage: String? = null
    private var mLackPermissionToDownloadMessage: String? = null

    private var mUserAgent: String? = null
    private var mUserAgentWithApplicationName: String? = null
    private val HTML_ENCODING = "UTF-8"
    private val HTML_MIME_TYPE = "text/html"
    private val JAVASCRIPT_INTERFACE = "ReactNativeWebView"
    private val HTTP_METHOD_POST = "POST"

    // Use `webView.loadUrl("about:blank")` to reliably reset the view
    // state and release page resources (including any running JavaScript).
    private val BLANK_URL = "about:blank"

    private val DEFAULT_DOWNLOADING_MESSAGE = "Downloading"
    private val DEFAULT_LACK_PERMISSION_TO_DOWNLOAD_MESSAGE =
        "Cannot download files as permission was denied. Please provide permission to write to storage, in order to download files."


    private var mContext: ReactApplicationContext = context

    fun createViewInstance(): RNCWebView? {
        val webView = RNCWebView(mContext)
        setupWebChromeClient(mContext, webView, mAllowsFullscreenVideo)
        mContext.addLifecycleEventListener(webView)
        mWebViewConfig.configWebView(webView)
        val settings = webView.settings
        settings.builtInZoomControls = true
        settings.displayZoomControls = false
        settings.domStorageEnabled = true
        settings.setSupportMultipleWindows(true)
        settings.allowFileAccess = false
        settings.allowContentAccess = false
        settings.allowFileAccessFromFileURLs = false
        setAllowUniversalAccessFromFileURLs(webView, false)
        setMixedContentMode(webView, "never")

        // Fixes broken full-screen modals/galleries due to body height being 0.
        webView.layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
        if (ReactBuildConfig.DEBUG) {
            WebView.setWebContentsDebuggingEnabled(true)
        }
        webView.setDownloadListener(DownloadListener { url, userAgent, contentDisposition, mimetype, contentLength ->
            webView.setIgnoreErrFailedForThisURL(url)
            val module = getModule() ?: return@DownloadListener
            val request: DownloadManager.Request = try {
                DownloadManager.Request(Uri.parse(url))
            } catch (e: IllegalArgumentException) {
                Log.w(TAG, "Unsupported URI, aborting download", e)
                return@DownloadListener
            }
            val fileName = URLUtil.guessFileName(url, contentDisposition, mimetype)
            val downloadMessage = "Downloading $fileName"

            //Attempt to add cookie, if it exists
            var urlObj: URL? = null
            try {
                urlObj = URL(url)
                val baseUrl = urlObj.protocol + "://" + urlObj.host
                val cookie = CookieManager.getInstance().getCookie(baseUrl)
                request.addRequestHeader("Cookie", cookie)
            } catch (e: MalformedURLException) {
                Log.w(TAG, "Error getting cookie for DownloadManager", e)
            }

            //Finish setting up request
            request.addRequestHeader("User-Agent", userAgent)
            request.setTitle(fileName)
            request.setDescription(downloadMessage)
            request.allowScanningByMediaScanner()
            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
            request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, fileName)
            module.setDownloadRequest(request)
            if (module.grantFileDownloaderPermissions(
                    getDownloadingMessageOrDefault(),
                    getLackPermissionToDownloadMessageOrDefault()
                )
            ) {
                module.downloadFile(
                    getDownloadingMessageOrDefault()
                )
            }
        })
        return webView
    }

    private fun setupWebChromeClient(
        reactContext: ReactApplicationContext,
        webView: RNCWebView,
        allowsFullscreenVideo: Boolean
    ) {
        val activity = reactContext.currentActivity
        if (allowsFullscreenVideo && activity != null) {
            val initialRequestedOrientation = activity.requestedOrientation
            val webChromeClient: RNCWebChromeClient =
                object : RNCWebChromeClient(reactContext, webView) {
                    override fun getDefaultVideoPoster(): Bitmap? {
                        return Bitmap.createBitmap(50, 50, Bitmap.Config.ARGB_8888)
                    }

                    override fun onShowCustomView(view: View, callback: CustomViewCallback) {
                        if (mVideoView != null) {
                            callback.onCustomViewHidden()
                            return
                        }
                        mVideoView = view
                        mCustomViewCallback = callback
                        activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
                        mVideoView.systemUiVisibility = FULLSCREEN_SYSTEM_UI_VISIBILITY
                        activity.window.setFlags(
                            WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
                            WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS
                        )
                        mVideoView.setBackgroundColor(Color.BLACK)

                        // Since RN's Modals interfere with the View hierarchy
                        // we will decide which View to hide if the hierarchy
                        // does not match (i.e., the WebView is within a Modal)
                        // NOTE: We could use `mWebView.getRootView()` instead of `getRootView()`
                        // but that breaks the Modal's styles and layout, so we need this to render
                        // in the main View hierarchy regardless
                        val rootView = rootView
                        rootView.addView(mVideoView, FULLSCREEN_LAYOUT_PARAMS)

                        // Different root views, we are in a Modal
                        if (rootView.rootView !== mWebView.rootView) {
                            mWebView.rootView.visibility = View.GONE
                        } else {
                            // Same view hierarchy (no Modal), just hide the WebView then
                            mWebView.visibility = View.GONE
                        }
                        mReactContext.addLifecycleEventListener(this)
                    }

                    override fun onHideCustomView() {
                        if (mVideoView == null) {
                            return
                        }

                        // Same logic as above
                        val rootView = rootView
                        if (rootView.rootView !== mWebView.rootView) {
                            mWebView.rootView.visibility = View.VISIBLE
                        } else {
                            // Same view hierarchy (no Modal)
                            mWebView.visibility = View.VISIBLE
                        }
                        activity.window.clearFlags(WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
                        rootView.removeView(mVideoView)
                        mCustomViewCallback.onCustomViewHidden()
                        mVideoView = null
                        mCustomViewCallback = null
                        activity.requestedOrientation = initialRequestedOrientation
                        mReactContext.removeLifecycleEventListener(this)
                    }
                }
            webView.webChromeClient = webChromeClient
        } else {
            var webChromeClient = webView.webChromeClient as RNCWebChromeClient?
            webChromeClient?.onHideCustomView()
            webChromeClient = object : RNCWebChromeClient(reactContext, webView) {
                override fun getDefaultVideoPoster(): Bitmap? {
                    return Bitmap.createBitmap(50, 50, Bitmap.Config.ARGB_8888)
                }
            }
            webView.webChromeClient = webChromeClient
        }
    }

    fun setUserAgent(view: WebView, userAgent: String?) {
        mUserAgent = userAgent
        setUserAgentString(view)
    }

    fun setApplicationNameForUserAgent(view: WebView, applicationName: String?) {
        when {
            applicationName != null -> {
                val defaultUserAgent = WebSettings.getDefaultUserAgent(view.context)
                mUserAgentWithApplicationName = "$defaultUserAgent $applicationName"
            }
            else -> {
                mUserAgentWithApplicationName = null
            }
        }
        setUserAgentString(view)
    }

    private fun setUserAgentString(view: WebView) {
        when {
            mUserAgent != null -> {
                view.settings.userAgentString = mUserAgent
            }
            mUserAgentWithApplicationName != null -> {
                view.settings.userAgentString = mUserAgentWithApplicationName
            }
            else -> {
                view.settings.userAgentString = WebSettings.getDefaultUserAgent(view.context)
            }
        }
    }

    fun setBasicAuthCredential(view: WebView, credential: ReadableMap?) {
        var basicAuthCredential: RNCBasicAuthCredential? = null
        if (credential != null) {
            if (credential.hasKey("username") && credential.hasKey("password")) {
                val username = credential.getString("username")
                val password = credential.getString("password")
                basicAuthCredential = RNCBasicAuthCredential(username, password)
            }
        }
        (view as RNCWebView).setBasicAuthCredential(basicAuthCredential)
    }

    fun getModule(): RNCWebViewModule? {
        return mContext.getNativeModule(RNCWebViewModule::class.java)
    }

    fun setMixedContentMode(view: WebView, mixedContentMode: String?) {
        if (mixedContentMode == null || "never" == mixedContentMode) {
            view.settings.mixedContentMode = WebSettings.MIXED_CONTENT_NEVER_ALLOW
        } else if ("always" == mixedContentMode) {
            view.settings.mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        } else if ("compatibility" == mixedContentMode) {
            view.settings.mixedContentMode = WebSettings.MIXED_CONTENT_COMPATIBILITY_MODE
        }
    }

    fun setAllowUniversalAccessFromFileURLs(view: WebView, allow: Boolean) {
        view.settings.allowUniversalAccessFromFileURLs = allow
    }

    private fun getDownloadingMessageOrDefault(): String? {
        return mDownloadingMessage ?: DEFAULT_DOWNLOADING_MESSAGE
    }

    private fun getLackPermissionToDownloadMessageOrDefault(): String? {
        return mLackPermissionToDownloadMessage
            ?: DEFAULT_LACK_PERMISSION_TO_DOWNLOAD_MESSAGE
    }

    fun setSource(view: RNCWebView, source: ReadableMap?) {
        if (source != null) {
            if (source.hasKey("html")) {
                val html = source.getString("html")
                val baseUrl = if (source.hasKey("baseUrl")) source.getString("baseUrl") else ""
                view.loadDataWithBaseURL(
                    baseUrl,
                    html!!,
                    HTML_MIME_TYPE,
                    HTML_ENCODING,
                    null
                )
                return
            }
            if (source.hasKey("uri")) {
                val url = source.getString("uri")
                val previousUrl = view.url
                if (previousUrl != null && previousUrl == url) {
                    return
                }
                if (source.hasKey("method")) {
                    val method = source.getString("method")
                    if (method.equals(HTTP_METHOD_POST, ignoreCase = true)) {
                        var postData: ByteArray? = null
                        if (source.hasKey("body")) {
                            val body = source.getString("body")
                            postData = try {
                                body!!.toByteArray(charset("UTF-8"))
                            } catch (e: UnsupportedEncodingException) {
                                body!!.toByteArray()
                            }
                        }
                        if (postData == null) {
                            postData = ByteArray(0)
                        }
                        view.postUrl(url!!, postData)
                        return
                    }
                }
                val headerMap = HashMap<String, String?>()
                if (source.hasKey("headers")) {
                    val headers = source.getMap("headers")
                    val iter = headers!!.keySetIterator()
                    while (iter.hasNextKey()) {
                        val key = iter.nextKey()
                        if ("user-agent" == key.lowercase(Locale.ENGLISH)) {
                            view.settings.userAgentString = headers.getString(key)
                        } else {
                            headerMap[key] = headers.getString(key)
                        }
                    }
                }
                view.loadUrl(url!!, headerMap)
                return
            }
        }
        view.loadUrl(BLANK_URL)
    }

    fun setMessagingModuleName(view: RNCWebView, value: String) {
        view.messagingModuleName = value
    }

    fun setCacheEnabled(view: RNCWebView, enabled: Boolean) {
        if (enabled) {
            val ctx: Context? = view.context
            if (ctx != null) {
                view.settings.setAppCachePath(ctx.cacheDir.absolutePath)
                view.settings.cacheMode = WebSettings.LOAD_DEFAULT
                view.settings.setAppCacheEnabled(true)
            }
        } else {
            view.settings.cacheMode = WebSettings.LOAD_NO_CACHE
            view.settings.setAppCacheEnabled(false)
        }
    }

    fun setIncognito(view: RNCWebView, enabled: Boolean) {
        // Don't do anything when incognito is disabled
        if (!enabled) {
            return;
        }

        // Remove all previous cookies
        CookieManager.getInstance().removeAllCookies(null);

        // Disable caching
        view.settings.cacheMode = WebSettings.LOAD_NO_CACHE
        view.settings.setAppCacheEnabled(false)
        view.clearHistory();
        view.clearCache(true);

        // No form data or autofill enabled
        view.clearFormData();
        view.settings.savePassword = false;
        view.settings.saveFormData = false;
    }

    fun setInjectedJavaScript(view: RNCWebView, injectedJavaScript: String) {
        view.injectedJS = injectedJavaScript
    }

    fun setInjectedJavaScriptBeforeContentLoaded(view: RNCWebView, value: String) {
        view.injectedJSBeforeContentLoaded = value
    }

    fun setInjectedJavaScriptForMainFrameOnly(view: RNCWebView, value: Boolean) {
        view.injectedJavaScriptForMainFrameOnly = value
    }

    fun setInjectedJavaScriptBeforeContentLoadedForMainFrameOnly(view: RNCWebView, value: Boolean) {
        view.injectedJavaScriptBeforeContentLoadedForMainFrameOnly = value
    }

    fun setJavaScriptCanOpenWindowsAutomatically(view: RNCWebView, value: Boolean) {
        view.settings.javaScriptCanOpenWindowsAutomatically = value
    }

    fun setShowsVerticalScrollIndicator(view: RNCWebView, value: Boolean) {
        view.isVerticalScrollBarEnabled = value
    }

    fun setShowsHorizontalScrollIndicator(view: RNCWebView, value: Boolean) {
        view.isHorizontalScrollBarEnabled = value
    }

    fun setMessagingEnabled(view: RNCWebView, value: Boolean) {
        view.setMessagingEnabled(value)
    }

    fun setMediaPlaybackRequiresUserAction(view: RNCWebView, value: Boolean) {
        view.settings.mediaPlaybackRequiresUserGesture = value
    }

    fun setHasScrollEvent(view: RNCWebView, value: Boolean) {
        view.setHasScrollEvent(value)
    }

    fun setJavaScriptEnabled(view: RNCWebView, enabled: Boolean) {
        view.settings.javaScriptEnabled = enabled
    }

}